### ISSUE TITLE
com.google.fonts/check/soft_dotted may be strict for now for GF

### ISSUE THREAD

**Issue by RosaWagner** [googlefonts/fontbakery#4085]
🔥 FAIL: Ensure soft_dotted characters lose their dot when combined with marks that replace the dot. ([com.google.fonts/check/soft_dotted](https://font-bakery.readthedocs.io/en/stable/fontbakery/profiles/universal.html#com.google.fonts/check/soft_dotted))
🔥 FAIL The dot of soft dotted characters used in orthographies must disappear in the following strings: į̀ į́ į̂ į̃ į̄ į̌ ị̀ ị́ ị̂ ị̃ ị̄
The dot of soft dotted characters should disappear in other cases, for example: į̆ į̇ į̈ į̉ į̊ į̋ į̏ į̑ į̒ į̵̀ į̵́ į̵̂ į̵̃ į̵̄ į̵̆ į̵̇ į̵̈ į̵̉ į̵̊ į̵̋ [code: soft-dotted]

We should bring a fix directly to gftools before making it a FAIL, at least for fonts not supporting `African` and `Beyond` glyphsets.

cc @moyogo @m4rc1e @vv-monsalve

**Comment by vv-monsalve** [comment]
> We should bring a fix directly to gftools before making it a FAIL, at least for fonts not supporting African and Beyond glyphsets.

Agree. Alternatively, it should receive only a  `WARN` if the font is not supporting African and Beyond

**Comment by colinmford** [comment]
I came here to make an issue about this as well — at the very least it should be a `WARN` instead of a `FAIL`

**Comment by m4rc1e** [comment]
Agreed. Make this a warn please.

**Comment by felipesanches** [comment]
@yanone said he wants to fix this one. If you plan on doing it, please reassign the issue to yourself.

**Comment by yanone** [comment]
Given that the African glyphsets are developed by Google, shall the check even remain in the `universal` profile after it's been loosened?

**Comment by yanone** [comment]
Another question:

[Ysabeau](https://raw.githubusercontent.com/CatharsisFonts/Ysabeau/master/sources/Ysabeau.glyphs) (still unreleased), has the following feature code:

```
lookup ccmp_Soft_Dotted {
	sub [idotbelow iogonek i-cy je-cy]' @CombiningTopAccents by [idotless_dotbelowcomb idotless_ogonekcomb idotless jdotless];
	sub [idotbelow iogonek i-cy je-cy]' @CombiningNonTopAccents @CombiningTopAccents by [idotless_dotbelowcomb idotless_ogonekcomb idotless jdotless];
} ccmp_Soft_Dotted;
```

Given that there are two Cyrillic glyphs in there, calculating African Latin support may not be sufficient to see whether the feature is needed. What are everyone’s thoughts on how to solve this?

Paging master @moyogo as well.

**Comment by moyogo** [comment]
@yanone @colinmford  @m4rc1e  The check catches Dutch j́, Ukrainian і́, or Lithuanian į̀. What specifically do you want the check not to FAIL for?

**Comment by moyogo** [comment]
@yanone hyperglot will report a language as supported when the required characters are in a font but their behaviour may be incomplete or incorrect, shaperglot should report a language as supported when the required characters seem to behave as expected. So in the case of checking the soft dotted behaviour, shaperglot has a chicken-or-egg problem.

**Comment by yanone** [comment]
@moyogo Thank you, that confirms my earlier understanding. My plan is now to simply calculate a font’s language coverage (as related to the soft-dotted-i feature) and add that to the output as additional information, but not downgrade the check result. In case of Dutch, the `jacute` will in many cases be already supported through Glyphs’ existing feature code, but that's exactly the learning curve that people will need to go through.

**Comment by m4rc1e** [comment]
My issue with this check is the amount of work and support it's going to cause. If a user implements GF Latin Core, do we expect Lithuanian, Navajo and Ukranian Latin to even be supported? I'm guessing we expect Lithuanian so this could render my argument invalid. However, even system fonts fail this,

<img width="727" alt="Screenshot 2023-05-22 at 10 33 25" src="https://github.com/googlefonts/fontbakery/assets/7525512/c0b419c2-ee2b-4552-a9ef-880e462d8d2d">

**Comment by moyogo** [comment]
@m4rc1e I have a ufo2ft filter that prepends the ccmp lookups to fix this to be used for gftools builder, but https://github.com/googlefonts/ufo2ft/pull/733 needs to be merged first.

**Comment by moyogo** [comment]
@m4rc1e:
- Dutch íj́ occurs in about 1 word per million but it usual form íj (which can be spelled íj́ when possible with the 1996 or 2006 spelling) occurs in about 5 words per million (500M words Dutch corpus), 
- Ukrainian і́ occurs in about 0.2 words per million in news sites (1M words Ukrainian 2020 news corpus), but occurs in about 34 words per million on the Ukrainian Wikipedia as it is used on the introduction line of articles to indicate word stress.
- Navajo į́ occurs in about 10925 words per million in Navajo Wikipedia (10K words Navajo corpus).
- Lithuanian į̇́ occurs in about 0 words per million (300K words Lithunian corpus), however reference works like the Academic Lithuanian Dictionary use it or use PUA characters for it so a better corpus is needed. For example the Academic Lithuanian Dictionary uses it but the Universal Lithuanian Encyclopedia uses a PUA character. There a reason PUA characters are used.

**Comment by RosaWagner** [comment]
For googlefonts profile, these features should follow the defined glyphsets:
- Dutch ij should be covered in **Latin_Core**
- Academic Lithuanian and Navajo should be part of the **Latin_Beyond** glyphset

- Ukrainian i acute… is it latin transliteration or actual i-cy ? First case it should be part of Beyond (although Latin_Core would already supports that in ccmp), second case is for cyrillic glyphset.

**Comment by m4rc1e** [comment]
I'd be interested in seeing how "íj́" renders across our collection. I have a feeling that for some families it will render correctly but it will fail the FB check. 

From the stats Denis and Rosalie have pulled up, I agree that it should work correctly for Dutch. The others are may a bit too esoteric.

**Comment by felipesanches** [comment]
@m4rc1e, these are the font files that FAIL the current implementation of **check/soft_dotted** on the GF Library today ([commit 98254375d5713d4de7c79ee3052281d4b3b7dce1](https://github.com/google/fonts/commit/98254375d5713d4de7c79ee3052281d4b3b7dce1)):

**Note:** This list was generated with `fontbakery check-googlefonts -c soft_dot ~/gfonts/ofl/f*/*.ttf -l FAIL | grep .ttf | sort | uniq | cut -d/ -f7` and a `"\n" => "\n* "` **replace all** operation on my text editor.

**Note 2:** FontBakery should have a feature for generating this kind of list, though...

* Fahkwang-BoldItalic.ttf
* Fahkwang-Bold.ttf
* Fahkwang-ExtraLightItalic.ttf
* Fahkwang-ExtraLight.ttf
* Fahkwang-Italic.ttf
* Fahkwang-LightItalic.ttf
* Fahkwang-Light.ttf
* Fahkwang-MediumItalic.ttf
* Fahkwang-Medium.ttf
* Fahkwang-Regular.ttf
* Fahkwang-SemiBoldItalic.ttf
* Fahkwang-SemiBold.ttf
* FamiljenGrotesk-Italic[wght].ttf
* FamiljenGrotesk[wght].ttf
* FanwoodText-Italic.ttf
* FanwoodText-Regular.ttf
* Farro-Bold.ttf
* Farro-Light.ttf
* Farro-Medium.ttf
* Farro-Regular.ttf
* Farsan-Regular.ttf
* FaunaOne-Regular.ttf
* Faustina-Italic[wght].ttf
* Faustina[wght].ttf
* Festive-Regular.ttf
* Figtree-Italic[wght].ttf
* Figtree[wght].ttf
* Finlandica-Italic[wght].ttf
* Finlandica[wght].ttf
* FiraCode[wght].ttf
* FiraMono-Bold.ttf
* FiraMono-Medium.ttf
* FiraMono-Regular.ttf
* FiraSansCondensed-BlackItalic.ttf
* FiraSansCondensed-Black.ttf
* FiraSansCondensed-BoldItalic.ttf
* FiraSansCondensed-Bold.ttf
* FiraSansCondensed-ExtraBoldItalic.ttf
* FiraSansCondensed-ExtraBold.ttf
* FiraSansCondensed-ExtraLightItalic.ttf
* FiraSansCondensed-ExtraLight.ttf
* FiraSansCondensed-Italic.ttf
* FiraSansCondensed-LightItalic.ttf
* FiraSansCondensed-Light.ttf
* FiraSansCondensed-MediumItalic.ttf
* FiraSansCondensed-Medium.ttf
* FiraSansCondensed-Regular.ttf
* FiraSansCondensed-SemiBoldItalic.ttf
* FiraSansCondensed-SemiBold.ttf
* FiraSansCondensed-ThinItalic.ttf
* FiraSansCondensed-Thin.ttf
* FiraSansExtraCondensed-BlackItalic.ttf
* FiraSansExtraCondensed-Black.ttf
* FiraSansExtraCondensed-BoldItalic.ttf
* FiraSansExtraCondensed-Bold.ttf
* FiraSansExtraCondensed-ExtraBoldItalic.ttf
* FiraSansExtraCondensed-ExtraBold.ttf
* FiraSansExtraCondensed-ExtraLightItalic.ttf
* FiraSansExtraCondensed-ExtraLight.ttf
* FiraSansExtraCondensed-Italic.ttf
* FiraSansExtraCondensed-LightItalic.ttf
* FiraSansExtraCondensed-Light.ttf
* FiraSansExtraCondensed-MediumItalic.ttf
* FiraSansExtraCondensed-Medium.ttf
* FiraSansExtraCondensed-Regular.ttf
* FiraSansExtraCondensed-SemiBoldItalic.ttf
* FiraSansExtraCondensed-SemiBold.ttf
* FiraSansExtraCondensed-ThinItalic.ttf
* FiraSansExtraCondensed-Thin.ttf
* FiraSans-BlackItalic.ttf
* FiraSans-Black.ttf
* FiraSans-BoldItalic.ttf
* FiraSans-Bold.ttf
* FiraSans-ExtraBoldItalic.ttf
* FiraSans-ExtraBold.ttf
* FiraSans-ExtraLightItalic.ttf
* FiraSans-ExtraLight.ttf
* FiraSans-Italic.ttf
* FiraSans-LightItalic.ttf
* FiraSans-Light.ttf
* FiraSans-MediumItalic.ttf
* FiraSans-Medium.ttf
* FiraSans-Regular.ttf
* FiraSans-SemiBoldItalic.ttf
* FiraSans-SemiBold.ttf
* FiraSans-ThinItalic.ttf
* FiraSans-Thin.ttf
* FjallaOne-Regular.ttf
* Foldit[wght].ttf
* Forum-Regular.ttf
* FragmentMono-Italic.ttf
* FragmentMono-Regular.ttf
* FrancoisOne-Regular.ttf
* Fraunces-Italic[SOFT,WONK,opsz,wght].ttf
* Fraunces[SOFT,WONK,opsz,wght].ttf
* Fredoka[wdth,wght].ttf
* Fruktur-Italic.ttf
* Fruktur-Regular.ttf
* Fuggles-Regular.ttf
* FuzzyBubbles-Bold.ttf
* FuzzyBubbles-Regular.ttf

**Comment by m4rc1e** [comment]
@felipesanches there are way more than this. Seems your results are only testing families which start with "F". Perhaps you haven't pasted them all.

**Comment by felipesanches** [comment]
exactly!

**Comment by m4rc1e** [comment]
Chuck in a count using `| wc -l`. I don't care so much about the family names, I want to know a percentage.

**Comment by m4rc1e** [comment]
1817 out of 3282 fonts in our collection currently fail this check.

**Comment by m4rc1e** [comment]
If I have a look at Whisper, we can see "íj́" renders correctly yet it fails the check because it probably doesn't soft dot the other combos well.

![s](https://github.com/googlefonts/fontbakery/assets/7525512/2e37f3cf-ccd9-4f3c-9c04-94fee0b0cda2)

**Comment by m4rc1e** [comment]
Same with Warnes, Markazi Text, Overpass Mono. There are loads of them. I feel this check is far too greedy.

**Comment by m4rc1e** [comment]
You're correct. I just skimmed the languages we support in [GF Latin Core](https://github.com/googlefonts/glyphsets/tree/main/GF_glyphsets/Latin#gf-latin-core) and Lithuanian is one of them, even though it has less than 5m speakers. I initially thought GF Latin Core didn't support it.

**Comment by m4rc1e** [comment]
I'd personally do the following for this check:

Fail if soft dotted combos don't work for GF Latin Core. Warn for soft dotted combos which don't work for all possible combos in the font.

I'd also move Lithuanian out of GF Latin Core since it has 2.6m speakers.

**Comment by moyogo** [comment]
@m4rc1e 
> I'd be interested in seeing how "íj́" renders across our collection. I have a feeling that for some families it will render correctly but it will fail the FB check.

For íj́ (with the combining acute character after j): from well to poorly depending on the fonts.
Some fonts do `sub iacute j' by j_acutecomb` or `sub iacute j by iacute_j.loclNLD` and do nothing with the acutecomb, leave it duplicated and mispositioned. Some fonts have not been updated to include acutecomb.

Note that the GoogleFonts Latin subset does not support Dutch íj́ (with the combining acute character after j) at all because of https://github.com/googlefonts/glyphsets/pull/102.

**Comment by m4rc1e** [comment]
Thanks Denis. I also did some digging and Maven Pro has a dedicated glyph

<img width="367" alt="Screenshot 2023-05-22 at 13 00 55" src="https://github.com/googlefonts/fontbakery/assets/7525512/6d69d1cc-e366-4b75-bf1d-5b1695def599">

Another worry of mine is that we're going to fail fonts based on not using the soft dotted approach when they do in fact have the correct appearance because they use whole glyph substitutions.

**Comment by m4rc1e** [comment]
> > Another worry of mine is that we're going to fail fonts based on not using the soft dotted approach when they do in fact have the correct appearance because they use whole glyph substitutions.
> 
> Maybe that can be automatically detected and accounted for in the check implementation?

Without using a very well trained OCR, I don't see how it's possible.

This issue is similar to my recommended contour count check. I'd love to be able to tell you that your glyph looks incorrect but I can only indicate that something may be wrong, hence why the check is a warn.

**Comment by RosaWagner** [comment]
The custom glyphs 'ijacute' is most of the time useless cause not part of any substitution feature, and anyway, it is better to have a proper ccmp feature for i and j — which we have been doing for a while now.

Lithuanian is supported by Latin Core because it is similar to Polish. The only difference is the Uogonek, but since Latin Core already supports ogonek, it didn't kill to have it to support all the main languages of Europe. But academic lithuanian is a special subset of Lithuanian, therefore should be covered by Beyond.

**Comment by moyogo** [comment]
@RosaWagner 
> Ukrainian i acute… is it latin transliteration or actual i-cy ?

This is the actual i-cy acutecomb: і́

Looking at a small sample of e-books, it seems to occur in about 2 words per million, which sits between the 0.2 words per million in the news sites corpus and the 34 words per million on the Ukrianian Wikipedia. Note that Latin í (with words in Cyrillic) is often used instead of Cyrillic і́, it’s not clear whether this is because of PDF-to-text conversion or not.

@m4rc1e @RosaWagner @felipesanches 
The check actually is going to silently pass as long as a substitution occurs and the soft-dotted character seem to lose their dot. It doesn’t check that the substitution looks correct 🫤

**Comment by m4rc1e** [comment]
Thanks Denis

> The check actually is going to silently pass as long as a substitution occurs and the soft-dotted character seem to lose their dot. It doesn’t check that the substitution looks correct 🫤

This is the part I don't like. If the results visually look fine, it shouldn't be a fail imo. I don't really care how they are constructed tbh.

**Comment by moyogo** [comment]
> This is the part I don't like. If the results visually look fine, it shouldn't be a fail imo. I don't really care how they are constructed tbh.

@m4rc1e Do you have an example?

**Comment by m4rc1e** [comment]
Cormorant

```
hb-shape ~/Type/fonts/ofl/cormorant/Cormorant\[wght\].ttf "íj́"
[iacute=0+275|j_acutecomb=1+255]
```

![corm](https://github.com/googlefonts/fontbakery/assets/7525512/f6dbfdee-38ff-49b0-b1f9-a86bd5612270)


```
>>> fontbakery check-googlefonts ~/Type/fonts/ofl/cormorant/Cormorant\[wght\].ttf -l FAIL -c dotted

The dot of soft dotted characters should disappear in other cases, for
          example: i̦̇ i̦̊ i̦̋ i̦̍ j̏ j̛̏ j̛̑ j̣̏ j̣̑ j̤̏ j̤̑ j̦̀ j̦́ j̦̃ j̦̄   
          j̦̆ j̦̇ j̦̈ j̦̉ j̦̊
```

**Comment by m4rc1e** [comment]
I'm going to be honest and just admit that these two families are the only ones in our collection ;-). However, I still don't care how people get to the end result. If it looks like an íj́, then it is an íj́. This also applies to other combos, if they don't use the dotlessi or dotlessj, I just don't care, as long as they look correct.

@moyogo what do you think about my [comment](https://github.com/googlefonts/fontbakery/issues/4085#issuecomment-1557067154)?

**Comment by moyogo** [comment]
@m4rc1e Nabla and Cormorant are not failing because of j́ or íj́ though. But I see what you mean.

Cormorant would fail at displaying languages that use its ɨ or ị with accents. In this case a fix would be to update the soft dotted lookup or prepending a more complete ccmp lookup.
Nabla, with the color and shadow, would fail for i and j with combining marks but also for other base letters with combining marks like g̃ or ỹ. In this case, a proper fix would be more than updating the lookup, although doing so would make the check pass.

**Comment by m4rc1e** [comment]
> Cormorant would fail at displaying languages that use its ɨ or ị with accents. 

Do any languages in GF Latin Core use these? Personally, I'd only want this check to fail for combinations which could exist in GF Latin Core.

**Comment by yanone** [comment]
I agree that in the best case the check shouldn't have to rely on specific `ccmp` code to be present in the font, but whole-glyph substitutions should pass as well. As I understand it, this should already be the case. The check compares the shaping buffer result against a [manually defined sequence](https://github.com/googlefonts/fontbakery/blob/main/Lib/fontbakery/profiles/universal.py#L1832) which uses unchanged base characters such as `i` and `j` plus marks. As long as the shaped buffer differs from that, the check passes for that base glyph. This implementation should be agnostic of how the single-glyph shaping comes about in the font.

However, it does not account for the plenty by-now historic implementations of the Dutch `íj́` as ligatures. The check only looks at single glyphs. So one problem with the check for sure is that it would have to involve visual comparisons here in case we have situations where only the Dutch `íj́` needs to be supported. Which brings me back to the logic of checking for languages or glyphsets.

I know that the issue topic here is to improve the check for GF, but it's currently sitting in the universal profile (where it should be), so I'd rather improve it in the most compatible way for everyone before we return to GF-specific implementations (probably involving glyphsets).

Since `vharfbuzz` is already used for the shaping in the check, the extra step towards a visual comparison via SVG->pixels might prove viable. Not OCR, just pixel comparison.

I find it not helpful to check for how many of our old fonts support the feature, because it’s explicitly a new feature that we want to support in the future where necessary.

If it turns out that the Dutch `íj́` is the _only_ sequence that needs to be supported in most cases, then the check should be adjusted to that because most fonts already support this I think.
However, if it turns out that next to Dutch most fonts also need to support other soft-dotted combinations, then we can skip the `íj́` part.

Is anyone able to write down in human language the conditions that we could use for narrowing down the _necessity_ part of our problem, both for universal profile and GF profile?

I try first:

**Universal Profile:**
It's unclear to universal Fontbakery which languages an author wants to fully support in a font, therefore I would leave the check as I've currently updated it incl. the language calculations for the author’s information. Foundries are always free to suppress the check if they don’t like it or downgrade its status. Personally I feel that this check is a very important addition to supporting minority languages and many of the discussions we’re leading here will very soon become obsolete once both `gftools` and Glyphs can support the feature out of the box, so I’m calling into question the need to fine-tune this check into becoming quite the monster (visual comparisons).

**GF Profile:**
Nail down which GF glyphsets require which combinations of the soft-dotted glyphs. If it turns out that Dutch is the only one for Latin Core, I'm willing to attempt a visual comparison and call it quits for Latin Core, while continuing with the check as-is (but with narrower glyph requirements) for the more extended GF glyphsets.

**Comment by yanone** [comment]
Note: I would add a comment to the current check’s output that the check only checks single glyphs so that authors can ignore it if their font already supports the Dutch `íj́` and they’re only targeting Dutch.

**Comment by yanone** [comment]
General information: The following languages are affected (True/False at the end of each line is whether Braah fully covers each one or not):

```
Western Frisian (Latn, 743,057 speakers) True
Lithuanian (Latn, 2,357,094 speakers) True
Navajo (Latn, 166,319 speakers) True
Dutch (Latn, 31,709,104 speakers) True
Basaa (Latn, 332,940 speakers) False
Igbo (Latn, 27,823,640 speakers) False
Belarusian (Cyrl, 10,064,517 speakers) False
Ukrainian (Cyrl, 29,273,587 speakers) False
Aghem (Latn, 38,843 speakers) False
```

So Ukrainian is only in Cyrillic. And Lithuanian won't even pass a 2.5M requirement, let alone 5M.
So Latin Core fonts (>5M speakers) really boils down to Dutch only.

**Comment by yanone** [comment]
So for the universal profile we could downgrade the status to WARN if languages <5M (or another number) speakers are affected for a font.

For GF profile we could simply do the same, since Latin Core will only yield Dutch with >5M. Then, if only Dutch is affected, the check should include checking for a ligature substitution.

**Comment by yanone** [comment]
Here’s an updated proposal for how to implement the check:

Have the supported speakers of a language defined as a config variable, default is 5M.
If a font fully covers a language that requires the soft-dotted feature and the font is missing the feature and speakers are over the threshold, FAIL, otherwise WARN.
If the result of the calculation yields only Dutch, attempt a visual comparison of the `íj́` and FAIL only if that fails.

This way the check wouldn't have to be reimplemented with different logic for each publisher and only the threshold config variable would be changed, if at all.

**Comment by simoncozens** [comment]
What strikes me about this check is that it's a specific check for a specific situation which comes up in language support for a few languages. In other words, you have the glyphset correct for a language, but more than just testing glyphset coverage, you want to check whether the font meets the language's expectations for how to use those glyphs within that language.

But that's precisely what shaperglot is designed for. Rather than enforcing a specific check for a specific set of glyphs, why not just hand the font to shaperglot and check that the covered languages are supported correctly?

**Comment by yanone** [comment]
Maybe here I would like to ask you to elaborate a bit on how `shaperglot` and the `gflanguages` work.

I’m assuming that the braces in the language definitions mean some sort of auxiliary status, even though `auxiliary` exists as a separate definition category. I’m assuming that because other language definitions (Aghem) include stuff like `{ɨ̀}` in the base defs which is a single letter, so the braces can’t signify ligatures.

The Dutch base definitions are `a á ä b c d e é ë f g h i í ï {ij} {íj́} j k l m n o ó ö p q r s t u ú ü v w x y z`, so they include `{íj́}`, but Braah (which supports Dutch according to shaperglot) only shapes `íj` correctly, which isn’t part of the definitions, but mis-shapes `íj́` which _is_ part of the definitions.

(This is all ignoring https://github.com/googlefonts/fontbakery/issues/4068, from which I understand that `íj` is the only stress that should be supported in fonts, not `íj́`. We'll return to that later, but ideally before the resolution of this issue here.)

![Bildschirmfoto 2023-05-23 um 15 57 08](https://github.com/googlefonts/fontbakery/assets/175386/f60edaa6-72b0-4a87-b614-f2376fb5672f)

For Navajo, which Braah also supports according to shaperglot, base definitions are `A Ą B C D E Ę G H I Į J K L Ł M N O Ǫ S T W X Y Z a á ą ą́ b c d e é ę ę́ g h i í į į́ j k l ł m n o ó ǫ ǫ́ s t w x y z ʼ`, which includes `į́` (without braces), but shaping is completely broken:

![Bildschirmfoto 2023-05-23 um 16 01 32](https://github.com/googlefonts/fontbakery/assets/175386/c53d5be0-092e-4ae3-9366-67f294978eea)

So how does shaperglot decide on language support?

**Comment by yanone** [comment]
Update: Sorry, I think I may have messed up this one because it appears to be an anchor issue, not a shaping issue. The Navajo glyph is actually composed correctly, and so is the `íj́`. The parts are just positioned incorrectly. 

(There are Fontbakery issue requests waiting in line related to anchors, which I think was because of this font)

**Comment by yanone** [comment]
@simoncozens Could you please elaborate a bit on how shaperglot works, maybe in comparison to how this check here works? That would help me get to the bottom of this.

Denis has manually defined a HB buffer for the unchanged sequence and compares the shaped buffer against it. If it differs, it counts as supported. But the GF language definitions don't contain an unshaped comparison (and why would they), so how does it work?

**Comment by simoncozens** [comment]
Shaperglot doesn't just check the glyphset support from the GF language definitions, although that is the first level of checking. The second level of checking uses its own language definition files to perform various shaping tests. One of those tests is `shaping_differs`, which is equivalent to what Denis has done here. Other test check for, for example, whether there are unattached marks left behind at the end of shaping (to make sure that all anchors correctly attach for a particular sequence), or whether activating/deactivating `locl` causes a buffer sequence to change (e.g., ensuring that a language-specific form of a glyph like eng is available).

It doesn't look like we actually have the right rules for Dutch at the moment, which is why "bad" fonts are passing. But I suggest that shaperglot, not fontbakery, is the right place for that knowledge.

**Comment by simoncozens** [comment]
I just added support for checking that there are no unattached marks when testing a given orthography, which means that Braah no longer supports Dutch or Navajo:

```
Font does not fully support language 'nl_Latn'
 * FAIL: Shaper didn't attach acutecomb to j
Font does not fully support language 'nv_Latn'
 * FAIL: Shaper didn't attach uni0328 to O
 * FAIL: Shaper didn't attach uni0328 to o
 * FAIL: Shaper didn't attach acutecomb to aogonek
 * FAIL: Shaper didn't attach acutecomb to iogonek
 * FAIL: Shaper didn't attach uni0328 to oacute
 * FAIL: Shaper didn't attach acutecomb to eogonek
```

**Comment by colinmford** [comment]
As someone who uses fontbakery but is not a part of GF, I agree with @yanone's proposal for the universal profile... I *really* like this check, I would definitely appreciate the additional info @yanone added in, I just think it should be a WARN.

**Comment by moyogo** [comment]
@yanone Sounds good to me. Thank you all!

**Comment by yanone** [comment]
Current output, incl. number of speakers and also added unsupported languages:

```
     WARN The dot of soft dotted characters used in                             
          orthographies must disappear in the following strings: i̊ i̋ j̀ j́ j̃ 
          j̄ j̈ ị̀                                                              
          ị́ ị̂ ị̃ ị̄                                                           
                                                                                
                                                                                
          The dot of soft dotted characters should disappear in other cases, for
          example: i̇ ǐ i̒ i̦̇ i̦̊ i̦̋ ǐ̦ i̦̒ j̆ j̇ j̉ j̊ j̋ ǰ j̒ j̦̀ j̦́ j̦̃
          j̦̄ j̦̆                                                               
                                                                                
          Your font fully covers the following languages that require the       
          soft-dotted feature: Navajo (Latn, 166,319 speakers), Western Frisian 
          (Latn, 743,057 speakers), Dutch (Latn, 31,709,104 speakers),          
          Lithuanian (Latn, 2,357,094 speakers).                                
                                                                                
          Your font does not cover the following languages that require the     
          soft-dotted feature: Ukrainian (Cyrl, 29,273,587 speakers), Aghem     
          (Latn, 38,843 speakers), Basaa (Latn, 332,940 speakers), Igbo (Latn,  
          27,823,640 speakers), Belarusian (Cyrl, 10,064,517 speakers). [code:  
          soft-dotted]                                                          


    Result: WARN
```

**Comment by tiroj** [comment]
I would say that if a font supports dynamic mark positioning, it should support soft-dotted i. This is generic character property handling, not something specific to individual language orthographies.

**Comment by RosaWagner** [comment]
@tiroj soft dotted i and j would be required wether we change the check or not. The question here is more about soft dotted idotbelow and iogonek when a font support only GF Latin Core, which doesn’t seem to be a necessity and make the onboarded process complicated.

**Comment by yanone** [comment]
This can be closed as well
